### PR TITLE
Validate addresses and report link errors

### LINKED ISSUE #240: Validate Application Link on creation/update
**Summary:** 

Currently, the AS returns 200 when creating/updating an application link with invalid data.

**Why do we need this?**

Current approach makes it difficult to inform the user that the linking failed when submitting the create/edit forms. We can detect this only when requesting the link stats, see below.

**What is already there? What do you see now?**

```
curl -v -H "Authorization: Bearer $access_token" -XPUT --data '{"link": {"api_key": "invalid", "network_server_address": "invalid"}, "field_mask": {"paths": ["api_key", "network_server_address"]}}' http://localhost:1885/api/v3/as/applications/test-app-id/link | jq

* TCP_NODELAY set
* Connected to localhost (::1) port 1885 (#0)
> PUT /api/v3/as/applications/test-app-id/link HTTP/1.1
> Host: localhost:1885
> User-Agent: curl/7.60.0
> Accept: */*
> Authorization: Bearer MFRWG.4NYFPAPKU47AKK3A4DXGLRWZQLK4RUZPZKFPOTQ.EIJK6L3KNDZTJGHJCCSOTUGTTY6VMOHNTD6IH42WDZXYEGLC73GA
> Content-Length: 133
> Content-Type: application/x-www-form-urlencoded
>
} [133 bytes data]
* upload completely sent off: 133 out of 133 bytes
< HTTP/1.1 200 OK
< Access-Control-Allow-Origin: *
< Content-Type: application/json
< Vary: Origin
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-Request-Id: 01D590HG4ANANQBDAK7HW1SEBS
< X-Xss-Protection: 1; mode=block
< Date: Wed, 06 Mar 2019 08:04:59 GMT
< Content-Length: 56
<
{
  "network_server_address": "invalid",
  "api_key": "invalid"
}

→ curl -v -H "Authorization: Bearer $access_token" http://localhost:1885/api/v3/as/applications/test-app-id/link/stats | jq

* TCP_NODELAY set
* Connected to localhost (::1) port 1885 (#0)
> GET /api/v3/as/applications/test-app-id/link/stats HTTP/1.1
> Host: localhost:1885
> User-Agent: curl/7.60.0
> Accept: */*
> Authorization: Bearer MFRWG.4NYFPAPKU47AKK3A4DXGLRWZQLK4RUZPZKFPOTQ.EIJK6L3KNDZTJGHJCCSOTUGTTY6VMOHNTD6IH42WDZXYEGLC73GA
>
< HTTP/1.1 404 Not Found
< Access-Control-Allow-Origin: *
< Content-Type: application/json
< Vary: Origin
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-Request-Id: 01D590JM9NDB8X78GTQXPA7TH4
< X-Xss-Protection: 1; mode=block
< Date: Wed, 06 Mar 2019 08:05:36 GMT
< Content-Length: 371
<

{
  "code": 5,
  "message": "error:pkg/applicationserver:not_linked (not linked to `test-app-id`)",
  "details": [
    {
      "@type": "type.googleapis.com/ttn.lorawan.v3.ErrorDetails",
      "namespace": "pkg/applicationserver",
      "name": "not_linked",
      "message_format": "not linked to `{application_uid}`",
      "attributes": {
        "application_uid": "test-app-id"
      },
      "correlation_id": "fa37c4ccf545412d991f82bb5d8e230f"
    }
  ]
}
```

**What is missing? What do you want to see?**

Error returned when linking the app if the linking failed.

**How do you propose to implement this?**

...

**What can you do yourself and what do you need help with?**

File the issue

cc @htdvisser

### PR DESCRIPTION
**Pull request by johanstokking**
**Summary:**

Closes #240 
References #252 

**Changes:**

- Add regex validation for addresses; host name with optional port, or blank. This includes the application link address. Note that `invalid` (as mentioned in #240) is still a valid host name
- Store link errors and report them when getting link statistics
- Get link from registry on each link task iteration to check if it still exists
- Unrelated fix of MQTT setup error handling in AS and GS not worth its own PR reported offline by @adriansmares

**Notes for Reviewers:**

@bafonins please check calls in #240 and see if this is what you need.

### CODE REVIEW

**Comment by bafonins** [comment]
Not sure if thats the behaviour we need.
```
→ curl -v -H "Authorization: Bearer $access_token" -XPUT --data '{"link": {"api_key": "invalid", "network_server_address": "<script>alert()</script>"}, "field_mask": {"paths": ["api_key", "network_server_address"]}}' http://localhost:1885/api/v3/as/applications/test-app-id/link | jq

> PUT /api/v3/as/applications/test-app-id/link HTTP/1.1
> Host: localhost:1885

< HTTP/1.1 200 OK

{
  "network_server_address": "<script>alert()</script>",
  "api_key": "invalid"
}
```

**Comment by johanstokking** [comment]
@bafonins that's a general validation issue that's open; https://github.com/TheThingsNetwork/lorawan-stack/issues/51

**Comment by bafonins** [comment]
> @bafonins please check calls in #240 and see if this is what you need.

This PR doesnt really close https://github.com/TheThingsNetwork/lorawan-stack/issues/240 .

**Comment by bafonins** [comment]
> If you want to get a result from Set() if linking actually succeeded, i.e. synchronously, then that gets quite complicated.

Thats it. I thought it is possible and would simplify things on the frontend.

Regarding the address regex - https://regex101.com/r/UwEYqI/1/.

**Comment by johanstokking** [comment]
@bafonins fixed the regex.

I see the simplification of things but it wouldn't be reliable enough. Realtime info will go through events, health and statistics through the endpoint. Link errors are now available there.

File: pkg/applicationserver/linking.go [hunk 1]
```diff
@@ -34,34 +34,41 @@ import (
...
 	as.StartTask(ctx, "link", func(ctx context.Context) error {
-		err := as.link(ctx, ids, target)
+		target, err := as.linkRegistry.Get(ctx, ids, []string{
+			"network_server_address",
+			"api_key",
+			"default_formatters",
+		})
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				log.FromContext(ctx).WithError(err).Error("Failed to get link")
```
Comment (johanstokking) [inline pkg/applicationserver/linking.go]: At this point, something really bad is going on, i.e. a registry failure. We may want to retry this.

If the link is not found, we return gracefully to stop the task, i.e. the link has been deleted.

**Comment by johanstokking** [comment]
@bafonis added `as.link.fail` event which can drive realtime update on link start/stop/fail indication, see also https://github.com/TheThingsNetwork/lorawan-stack/issues/28.

**Pull request merged**